### PR TITLE
Allow tunring off DBCSR ACC with env variable

### PR DESCRIPTION
**Pull request by abussy**
This PR allows turning off DBCSR GPU acceleration at initialization, even if the library is compiled with the `-D__DBCSR_ACC` flag. This can be done by setting the environment variable `DBCSR_TURN_OFF_ACC=1`, or by passing a logical to the `dbcsr_set_config` subroutine (this will be useful for a future CP2K keyword).

Most changes take place in the `dbcsr_config.F` file. There, a new function called `use_acc()` is defined. It returns a logical, and replaces the `has_acc` variable used in the rest of the code.

The PR was validated by successfully running the CP2K regtests on a GPU machine, with and without the `DBCSR_TURN_OFF_ACC=1` environment variable.

Generally, the ability of turning off GPU acceleration at runtime should help with debugging, and with edge cases where DBCSR ACC under-performs (see #795).

### CODE REVIEW

**Comment by hfp** [comment]
With the environment variable `DBCSR_N_STACKS=0` existing already, does it make this PR superfluous or do you still want it?

**Comment by alazzaro** [comment]
We still need it since we still move data to the gpu and run transpose kernel.

Thanks @abussy , I will review the PR next week.

**Comment by abussy** [comment]
Gentle ping to review this PR :) It would be nice if it makes it to CP2K's next release

**Comment by alazzaro** [comment]
thanks for the PR.
I will start the review, but first let me put here some general comments for my understanding of the proposed changes:
1. we are replacing `has_acc`[ boolean parameter](https://github.com/cp2k/dbcsr/blob/01e10a076e5346704864625c348e4992c3383cf9/src/core/dbcsr_config.F#L64) with `use_acc` function
2. `has_acc` is used within the use_acc and when setting the default for [mm_dense](https://github.com/cp2k/dbcsr/blob/01e10a076e5346704864625c348e4992c3383cf9/src/core/dbcsr_config.F#L154)
3. We use the new conf flag `TURN_OFF_ACC` to disable the GPU when DBCSR is compiled for GPU

OK, the first "logical" problem is `TURN_OFF_ACC` has to be `TRUE` to disable the GPU. I would prefer the opposite meaning, so we can use `RUN_ON_GPU` so it is TRUE by default whenever has_acc is TRUE and people can set to FALSE if you don't want to run on the GPU (we are not really "turning off" the GPU, simply we are not running on it...).

In terms of output, I will keep has_acc when we output the [configuration](https://github.com/cp2k/dbcsr/blob/01e10a076e5346704864625c348e4992c3383cf9/src/core/dbcsr_config.F#L610).

The last problem is that it must forbidden to change this value multiple times within a DBCSR run, otherwise will get configuration problems (we are replacing a macro at compile time with something at runtime). This functionality is not available in the library, I need to think about (in any case this is not relevant to your PR).

**Comment by abussy** [comment]
Sure, using a positive boolean like `RUN_ON_GPU` makes sense, I agree.

Concerning the danger of changing this value multiple times during a run, I think it should be safe as it is. The `use_acc()` function is defined during the DBCSR configuration step.

File: src/core/dbcsr_config.F [hunk 1]
```diff
@@ -607,7 +625,12 @@ SUBROUTINE dbcsr_print_config(unit_nr)
          END IF
       END BLOCK
 
-      IF (has_acc) THEN
+      IF (dbcsr_cfg%turn_off_acc%val) THEN
+         WRITE (UNIT=unit_nr, FMT='(1X,A,T81,A4)') &
+            "DBCSR| ACC is turned off: only CPU is used", dbcsr_cfg%turn_off_acc%print_source()
+      END IF
+
+      IF (use_acc()) THEN
```
Comment (alazzaro) [inline src/core/dbcsr_config.F:633]: We should keep has_acc here, this is because people compiled for the GPU (even if they are not running on it).

**Comment by alazzaro** [comment]
> Concerning the danger of changing this value multiple times during a run, I think it should be safe as it is. The `use_acc()` function is defined during the DBCSR configuration step.

The [dbcsr_set_config](https://github.com/cp2k/dbcsr/blob/01e10a076e5346704864625c348e4992c3383cf9/src/dbcsr_api.F#L174) is a public function, so people can call it after the init. Actually, this is what CP2K does. That means we can call it once with GPU_RUN disabled, then run a multiplication, then call set_config and set GPU_RUN to true and at this point I'm not sure what it will happen... We need to protect for this case, as said this is not part of this PR (I will open an issue as a reminder).

**Comment by abussy** [comment]
I implemented the vast  majority of the suggested changes:
- the environment variable is now called `RUN_ON_GPU`
- `has_acc` is used to control the printing of ACC paramters, and not `use_acc()`
- the information about whether GPU support is enabled or not is printed with the rest of ACC info
- no more hardcoded CPU defaults (e.g. stack_size of 1000) with the introduction of new paramters variables

File: src/core/dbcsr_config.F [hunk 2]
```diff
@@ -419,9 +423,21 @@ SUBROUTINE dbcsr_set_config( &
...
       CALL dbcsr_cfg%n_stacks%set(nstacks)
-      CALL dbcsr_cfg%mm_stack_size%set(mm_stack_size)
       CALL dbcsr_cfg%mm_driver%set(mm_driver)
 
+      ! If ACC is turned-off, use the CPU defaults
+      IF (.NOT. PRESENT(mm_stack_size) .AND. dbcsr_cfg%turn_off_acc%val) THEN
+         CALL dbcsr_cfg%mm_stack_size%set(1000)
+      ELSE
+         CALL dbcsr_cfg%mm_stack_size%set(mm_stack_size)
+      END IF
```
[the remark this answers is not in the document]
Comment (abussy) [reply src/core/dbcsr_config.F]: And resetting the value to the CPU default (unless another value is requested by the user) makes sense, as one expects the full CPU behaviour when disabling GPU support

**Pull request merged**